src/consumer.rs

```diff
@@ -573,35 +583,75 @@ impl<Exe: Executor + ?Sized> ConsumerEngine<Exe> {
+                    self.send_to_consumer(id, payload, now).await?;
                 }
             }
-            None => {
-                self.tx
-                    .send(Ok((message.message_id.clone(), payload)))
-                    .await
-                    .map_err(|e| {
-                        error!("tx returned {:?}", e);
-                        Error::Custom("tx closed".to_string())
-                    })?;
-                if let Some(duration) = self.unacked_message_redelivery_delay {
-                    //info!("inserting id {:?} to trigger unack at {:?}",
-                    //      message.message_id, Instant::now()+duration);
-                    self.unacked_messages
-                        .insert(message.message_id, Instant::now() + duration);
+            None => match (message.redelivery_count, self.dead_letter_policy.as_ref()) {
+                (Some(redelivery_count), Some(dead_letter_policy)) => {
+                    // Send message to Dead Letter Topic and ack message in original topic
+                    if redelivery_count as usize >= dead_letter_policy.max_redeliver_count {
+                        self.client
+                            .send(
+                                dead_letter_policy.dead_letter_topic.clone(),
+                                payload.data,
+                                producer::ProducerOptions::default(),
+                            )
+                            .await?
+                            .await
+                            .map_err(|e| {
+                                error!("One shot cancelled {:?}", e);
+                                Error::Custom("DLQ send error".to_string())
+                            })?;
+
+                        self.ack_tx
+                            .send(AckMessage::Ack(
+                                MessageData {
+                                    id: message.message_id,
+                                    batch_size: None,
+                                },
+                                false,
+                            ))
+                            .await
+                            .map_err(|e| {
+                                error!("ack tx returned {:?}", e);
+                                Error::Custom("tx closed".to_string())
+                            })?;
+                    } else {
+                        self.send_to_consumer(message.message_id, payload, Instant::now())
+                            .await?
+                    }
                 }
-            }
+                _ => {
+                    self.send_to_consumer(message.message_id, payload, Instant::now())
+                        .await?
+                }
+            },
         }
 
         Ok(())
     }
 
+    async fn send_to_consumer(
+        &mut self,
+        message_id: MessageIdData,
+        payload: Payload,
+        request_time: Instant,
```

Comment:
This would probably be easier to read initialized in the function rather than as an arg, since it's always `Instant::now()` (and would be confusing to be anything else)

Reply:
I agree. The only reason why I left the initialization outside the function was because of the batched message use case, where it was previously doing `now = Instant::now()` and then using the same value for all iterations of the loop. [link](https://github.com/wyyerd/pulsar-rs/blob/master/src/consumer.rs#L574)

Do you think its ok for the batched message to have slightly different values in the `unacked_messages` map?

Reply:
Hm, I think ideally the whole batch would go off of the timestamp of the _last_ message, but there's no way to do that. Probably having them have different timestamps then makes the most sense, because it at least prevents the case of a large batch and a slow consumer causing the timeout to be reached before the last message has been received.